### DISCUSSION TITLE
Check Field No. / Field Type in Tables used with Transferfields

### DISCUSSION THREAD

**Discussion by JohnnyUndercover** [StefanMaron/BusinessCentral.LinterCop#137] [New Rule]
In BC Transferfields is used to move data from one table to another. E.g. when posting a document data is moved from Sales Header table to Sales Invoice Header table. Transferfields uses the field numbers to map the fields between source and target table. if a field with the same no. but diffrent datatype is created in source and target transferfields will throw an error. To prevent this behauvior, when adding a field to a table the same field should be created in all tables where the table is used as the source for transferfields.

For example if you create a tableextension for the Sales Header table and add a field (50000; "My Field"; Code[20]) the same field should be created in the tables "Sales Shipment Header", "Sales Invoice Header", "Sales Cr.Memo Header", "Sales Header Archive", "Return Receipt Header". 

I would love to get a error when i create a field in a table thats used with transferfields and the field no is already used by another table or a warning when i forget to create the field in all connected tables.

I don't know if it is possible to check this completle dynamic. For me it would be enought to have it checkt for the follwing combinations:

**Sales Document Headers**
Table	36	Sales Header
Table	110	Sales Shipment Header
Table	112	Sales Invoice Header
Table	114	Sales Cr.Memo Header
Table	5107	Sales Header Archive
Table	6660	Return Receipt Header

**Sales Document Lines**
Table	37	Sales Line
Table	111	Sales Shipment Line
Table	113	Sales Invoice Line
Table	115	Sales Cr.Memo Line
Table	5108	Sales Line Archive
Table	6661	Return Receipt Line

**Purchase Document Headers**
Table	38	Purchase Header
Table	120	Purch. Rcpt. Header
Table	122	Purch. Inv. Header
Table	124	Purch. Cr. Memo Hdr.
Table	5109	Purchase Header Archive
Table	6650	Return Shipment Header

**Purchase Document Lines**
Table	39	Purchase Line
Table	121	Purch. Rcpt. Line
Table	123	Purch. Inv. Line
Table	125	Purch. Cr. Memo Line
Table	5110	Purchase Line Archive
Table	6651	Return Shipment Line

**Contact realted Tables**
Table	18	Customer
Table	23	Vendor
Table	270	Bank Account
Table	5050	Contact

Btw. thanks for your effort and work creating this linter 👍

**Comment by StefanMaron** [comment]
Thanks for the idea and the detailed description!

In general I like would also like to see a rule like this. But I am not sure how this can be solved.  
I dont like to hard code some tables. But on the other hand I dont think the linter should parse all the code of all dependencies for transferfields.

I could imagine a rule that warns on the usage of transferfields, but then again, you never know what kind of fields get introduced by table extensions you dont know about yet.

I am not sure if transferfields is still the way to go because of those concerns.

**Reply by JohnnyUndercover** [reply]
I totally agree, transferfields in its current state is not the way to go. Nevertheless microsoft is still using it.

Personally i think having it hardcoded would be better than nothing. Because if you miss a field and your extension gets published in most cases posting documents ist blocket until you published a new version of your app.
But i do understand your concerns implementing it like this because it could potentially break if microsoft is changing something and having something hardcoded is always ugly.

Mybe the best solution would be some test codeunits. But then again, some hint at tables like sales header would be very usefull to know if this table is used as a source for transferfields. Maybe it is easier to just get the info that transferfields is used instead of checking every field.

**Comment by rvanbekkum** [comment]
I saw that the "AL Toolbox" extension has some commands that copy fields to related tables and pages: https://github.com/BartPermentier/al-toolbox#action-create-related-tables
I haven't used it myself yet though, but I saw that the configurations of what is related is stored in a JSON object.
Hardcoded in that extension though, but I think you could have it in a separate .json file (e.g., LinterCop.json, is the first that comes to mind) if you wanted to have it configurable.

**Comment by pri-kise** [comment]
Don't forget that there are even more tables that use TransferFields currently.
e.g. Warehouse Activity Header/Line, TimeSheet, Production ....
The CompanialCop seems to have something like this implemented although there are still some tables missing.
https://github.com/Companial/CompanialCop/blob/main/Design/Rule0027AnalyzeTransferField.cs
https://github.com/Companial/CompanialCop/issues/1

**Reply by Arthurvdv** [reply]
The [CompanialCop](https://github.com/Companial/CompanialCop) has indeed already created an extensive rule for this: ```CM0027 Tables coupled with TransferFields must have matching fields.```
 
There was a request on X/Twitter for @tinestaric to bring this rule over to the LinterCop, so with some luck in one of the upcoming releases this will be added to the LinterCop.

@tinestaric: Maybe we should discuss our approach on adding possible missing tables? Maintaining this rule on both CodeCops could create some extra overhead? 🤔

**Reply by tinestaric** [reply]
@Arthurvdv I think I'll find some time in the next two weeks, and I can create a preliminary PR, and we can then refine it. 
As for maintaining the rule and adding missing pairs, I think LinterCop should be the central place. The community is here, and they can help spot more missing tables. I think it's the better approach, that we all contribute to LinterCop.

Should we improve anything in CompanialCop, I'll make suggestions to move it here, but likewise, it's on us, to move any improvements from LinterCop to our repo, if we want to continue maintaining a separate copy.

**Reply by pri-kise** [reply]
Maybe some kind of setting might be useful to add custom Tables to this rule in the future.
I know some code, where TransferField is used to populate custom tables

**Reply by tinestaric** [reply]
The rule does analyze TransferField invocations for custom tables. If it finds a TransferFields(MyTable1, MyTable2) or TransferFields(Customer, MyTable1) it will compare the fields of these pairs. 

However, a setting would be useful if you'd like to add table pairs from a dependency using TransferFields. These are the ones the analyzer cannot get to and doesn't know it should compare more table pairs.

**Reply by tinestaric** [reply]
@pri-kise @Arthurvdv 
The PR is up: https://github.com/StefanMaron/BusinessCentral.LinterCop/pull/429

I went through the base app again, I found plenty more table pairs this time around. Though there might still be some missing.
